core/tests/data/ser/json.cpp

```diff
@@ -0,0 +1,93 @@
+#include <doctest/doctest.h>
+#include <glm/vec3.hpp>
+
+#include <cubos/core/data/ser/debug.hpp>
+#include <cubos/core/data/ser/json.hpp>
+#include <cubos/core/reflection/external/glm.hpp>
+#include <cubos/core/reflection/external/map.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+#include <cubos/core/reflection/external/uuid.hpp>
+#include <cubos/core/reflection/external/vector.hpp>
+#include <cubos/core/reflection/traits/constructible_utils.hpp>
+#include <cubos/core/reflection/traits/fields.hpp>
+
+using cubos::core::data::JSONSerializer;
+using cubos::core::reflection::FieldsTrait;
+using cubos::core::reflection::Type;
+
+#define AUTO_TEST(type, value, expected)                                                                               \
+    ser.write<type>(value);                                                                                            \
+    json = ser.output();                                                                                               \
+    CHECK((json.dump()) == (expected));
+
+TEST_CASE("data::JSONSerializer")
+{
+    JSONSerializer ser{};
+    auto json = ser.output();
+
+    /// [Primitive types]
+    AUTO_TEST(bool, true, "true");
+    AUTO_TEST(char, 'a', "97");
+    AUTO_TEST(int8_t, -128, "-128");
+    AUTO_TEST(int16_t, -32768, "-32768");
+    AUTO_TEST(int32_t, -2147483648, "-2147483648");
+    AUTO_TEST(int64_t, -9223372036854775807, "-9223372036854775807");
+    AUTO_TEST(uint8_t, 255, "255");
+    AUTO_TEST(uint16_t, 65535, "65535");
+    AUTO_TEST(uint32_t, 4294967295, "4294967295");
+    AUTO_TEST(uint64_t, 18446744073709551615ULL, "18446744073709551615");
+    AUTO_TEST(float, 1.5F, "1.5");
+    AUTO_TEST(double, 1.00000000000005, "1.00000000000005");
+
+    /// [String]
+    AUTO_TEST(std::string, "Hello, world!", "\"Hello, world!\"");
+
+    /// [UUID]
+    AUTO_TEST(uuids::uuid, *uuids::uuid::from_string("f7063cd4-de44-47b5-b0a9-f0e7a558c9e5"),
+              "\"f7063cd4-de44-47b5-b0a9-f0e7a558c9e5\"");
+
+    /// [Map]
+    std::map<std::string, bool> map{{"false", false}, {"true", true}};
+    AUTO_TEST(decltype(map), map, "{\"false\":false,\"true\":true}");
```

Comment:
You should also cover a map which doesn't have strings as keys, I suspect they will fail. E.g., `std::map<int, std::string>`. In JSON, you can't have a key of a structured type, unlike values. All keys must be strings. Thus, you will have to handle it specially. For example, in the JSON deserializer, I stored an extra field in the deserializer class, `mIsKey`. On hooks, if the `mIsKey` flag is set to true, instead of storing into `mJSON`, you would store into a `mKey` string variable. Then, on the dictionary trait, before calling write for the key, you would set `mIsKey` to true, and afterwards extract the result from the `mKey` variable, isntead of `mJSON`.